File: sdks/python/src/fuse/_services/snapshots.py
from __future__ import annotations

from typing import Optional
from urllib.parse import quote

from .._transport import Transport
from ..types import Snapshot, SnapshotPage, SnapshotRequest

# the server's max page size; list() requests this internally so it walks
# every page in as few round trips as possible.
_MAX_PAGE_LIMIT = 200


class SnapshotsService:
    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def create(
        self, vm_id: str, request: Optional[SnapshotRequest] = None
    ) -> Snapshot:
        # checkpoints a running environment. with no request (or with live
        # unset) this takes a disk snapshot, the rootfs and nothing else; pass
        # SnapshotRequest(live=True) to capture the guest's memory and vcpu
        # state as well.
        #
        # the returned Snapshot.kind reports what was actually written, which
        # can be "disk" even for a live request that reached a host too old to
        # honour it. a caller that depends on the memory being there should
        # check it. live against a backend with no live-snapshot support raises
        # ApiError with status 501, not a conflict.
        if not vm_id:
            raise ValueError("vm id is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}/snapshots"
        resp = self._t.request("POST", path, body=request or SnapshotRequest())
        return Snapshot.model_validate(resp.json())

    def resolve(self, layer_key: str, arch: str) -> Optional[Snapshot]:
        # resolves one layer cache key and architecture to the newest ready
        # build artifact, or None when there is none.
        #
        # a miss is not an error and never raises: a cold cache is the normal
        # state of a first build, and modelling it as a failure would make every
        # caller wrap this in a try/except to discover nothing was wrong.
        #
        # arch is required rather than defaulted. an ext4 rootfs is not portable
        # across architectures, so resolving without one could hand back an
        # artifact the caller cannot boot, at the exact moment it believes it
        # got a hit.
        #
        # the scope searched comes from how the client authenticated; there is
        # deliberately no tenant parameter.
        if not layer_key:
            raise ValueError("layer key is required")
        if not arch:
            raise ValueError("arch is required")
        resp = self._t.request(
            "GET",
            "/v1/snapshots/resolve",
            params={"layer_key": layer_key, "arch": arch},
        )
        body = resp.json()
        if not body.get("found") or not body.get("snapshot"):
            return None
        return Snapshot.model_validate(body["snapshot"])

    def list(
        self,
        *,
        vm_id: str = "",
        task_id: str = "",
        tenant_id: str = "",
        state: str = "",
        layer_key: str = "",
        arch: str = "",
        cursor: str = "",
    ) -> list[Snapshot]:
        # returns every snapshot matching the filters, transparently walking
        # every result page. for explicit single-page control (e.g. a cursor
        # from a previous call), use list_page.
        out: list[Snapshot] = []
        while True:
            page = self.list_page(
                vm_id=vm_id,
                task_id=task_id,
                tenant_id=tenant_id,
                state=state,
                layer_key=layer_key,
                arch=arch,
                limit=_MAX_PAGE_LIMIT,
                cursor=cursor,
            )
            out.extend(page.snapshots)
            if not page.next_cursor:
                break
            cursor = page.next_cursor
        return out

    def list_page(
        self,
        *,
        vm_id: str = "",
        task_id: str = "",
        tenant_id: str = "",
        state: str = "",
        layer_key: str = "",
        arch: str = "",
        limit: int = 0,
        cursor: str = "",
    ) -> SnapshotPage:
        # returns one page of snapshots matching the filters.
        #
        # layer_key narrows to the build layers taken after one setup step,
        # and arch to the artifacts built on one architecture ("amd64",
        # "arm64"). arch is a separate filter rather than part of the layer
        # key because a rootfs is not portable across architectures, so a
        # layer lookup that does not constrain arch can be served bytes it
        # cannot boot. an empty filter is dropped by clean_params, so it means
        # "do not filter" rather than "match artifacts with no layer key".
        params: dict[str, str] = {
            "vm_id": vm_id,
            "task_id": task_id,
            "tenant_id": tenant_id,
            "state": state,
            "layer_key": layer_key,
            "arch": arch,
        }
        if limit > 0:
            params["limit"] = str(limit)
        if cursor:
            params["cursor"] = cursor
        resp = self._t.request("GET", "/v1/snapshots", params=params)
        return SnapshotPage.model_validate(resp.json())

    def get(self, snapshot_id: str) -> Snapshot:
        if not snapshot_id:
            raise ValueError("snapshot id is required")
        resp = self._t.request("GET", f"/v1/snapshots/{quote(snapshot_id, safe='')}")
        return Snapshot.model_validate(resp.json())

    def delete(self, snapshot_id: str) -> None:
        if not snapshot_id:
            raise ValueError("snapshot id is required")
        self._t.request("DELETE", f"/v1/snapshots/{quote(snapshot_id, safe='')}")

    def restore(self, snapshot_id: str) -> None:
        if not snapshot_id:
            raise ValueError("snapshot id is required")
        path = f"/v1/snapshots/{quote(snapshot_id, safe='')}"
        self._t.request("POST", path, params={"action": "restore"})

File: sdks/rust/src/snapshots.rs
use std::sync::Arc;

use reqwest::Method;
use serde::Deserialize;

use crate::error::Error;
use crate::transport::{pagination, require, Transport, MAX_PAGE_LIMIT};
use crate::types::{Arch, Snapshot, SnapshotMode, SnapshotRequest, SnapshotState};

/// Filters and pagination for [`Snapshots::list`] and
/// [`Snapshots::list_page`].
#[derive(Debug, Clone, Default)]
#[non_exhaustive]
pub struct ListSnapshotsOptions {
    pub vm_id: Option<String>,
    pub task_id: Option<String>,
    pub tenant_id: Option<String>,
    pub state: Option<SnapshotState>,
    /// Narrows to one creation mode, e.g. [`SnapshotMode::Build`] for build
    /// artifacts.
    pub mode: Option<SnapshotMode>,
    /// Matches the snapshot's metadata name, the lookup key `fuse build`
    /// sets on an artifact.
    pub name: Option<String>,
    /// Narrows to the build layers taken after one setup step. `None` means
    /// "do not filter", never "match artifacts with no layer key".
    pub layer_key: Option<String>,
    /// Narrows to artifacts built on one architecture. It is a separate
    /// filter from `layer_key` rather than part of it because a rootfs is
    /// not portable across architectures, so a layer lookup that does not
    /// constrain arch can be served bytes it cannot boot.
    pub arch: Option<Arch>,
    /// Caps the page size (server default 50, max 200). Only consulted by
    /// [`Snapshots::list_page`] — [`Snapshots::list`] always requests the
    /// server's max page size internally since it walks every page.
    pub limit: Option<u32>,
    /// Resumes after a previous page's [`SnapshotPage::next_cursor`]. It is
    /// an opaque token; treat it as such rather than parsing it.
    pub cursor: Option<String>,
}

impl ListSnapshotsOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn vm_id(mut self, vm_id: impl Into<String>) -> Self {
        self.vm_id = Some(vm_id.into());
        self
    }

    pub fn task_id(mut self, task_id: impl Into<String>) -> Self {
        self.task_id = Some(task_id.into());
        self
    }

    pub fn tenant_id(mut self, tenant_id: impl Into<String>) -> Self {
        self.tenant_id = Some(tenant_id.into());
        self
    }

    pub fn state(mut self, state: SnapshotState) -> Self {
        self.state = Some(state);
        self
    }

    pub fn mode(mut self, mode: SnapshotMode) -> Self {
        self.mode = Some(mode);
        self
    }

    pub fn name(mut self, name: impl Into<String>) -> Self {
        self.name = Some(name.into());
        self
    }

    pub fn layer_key(mut self, layer_key: impl Into<String>) -> Self {
        self.layer_key = Some(layer_key.into());
        self
    }

    pub fn arch(mut self, arch: Arch) -> Self {
        self.arch = Some(arch);
        self
    }

    pub fn limit(mut self, limit: u32) -> Self {
        self.limit = Some(limit);
        self
    }

    pub fn cursor(mut self, cursor: impl Into<String>) -> Self {
        self.cursor = Some(cursor.into());
        self
    }
}

/// One page of a list call, plus the cursor to fetch the next one.
#[derive(Debug, Clone, Default, Deserialize)]
pub struct SnapshotPage {
    #[serde(default)]
    pub snapshots: Vec<Snapshot>,
    /// `None` once there are no more results.
    #[serde(default)]
    pub next_cursor: Option<String>,
}

/// The snapshots service, from
/// [`Client::snapshots`](crate::Client::snapshots).
#[derive(Clone)]
pub struct Snapshots {
    t: Arc<Transport>,
}

impl Snapshots {
    pub(crate) fn new(t: Arc<Transport>) -> Self {
        Self { t }
    }

    /// Takes a snapshot of a running environment.
    pub async fn create(&self, vm_id: &str, request: SnapshotRequest) -> Result<Snapshot, Error> {
        require(vm_id, "vm id")?;
        let request = self
            .t
            .request(Method::POST, &["v1", "environments", vm_id, "snapshots"])
            .json(&request);
        self.t.send_json(request).await
    }

    /// Returns every snapshot matching `options`, transparently walking
    /// every result page. For explicit single-page control (e.g. a CLI
    /// `--cursor` flag), use [`Snapshots::list_page`].
    pub async fn list(&self, options: ListSnapshotsOptions) -> Result<Vec<Snapshot>, Error> {
        let mut out = Vec::new();
        let mut page_options = options;
        page_options.limit = Some(MAX_PAGE_LIMIT);
        loop {
            let page = self.list_page(page_options.clone()).await?;
            out.extend(page.snapshots);
            match page.next_cursor {
                Some(cursor) if !cursor.is_empty() => page_options.cursor = Some(cursor),
                _ => return Ok(out),
            }
        }
    }

    /// Returns one page of snapshots matching `options`.
    pub async fn list_page(&self, options: ListSnapshotsOptions) -> Result<SnapshotPage, Error> {
        let mut query: Vec<(&str, String)> = Vec::new();
        if let Some(vm_id) = &options.vm_id {
            query.push(("vm_id", vm_id.clone()));
        }
        if let Some(task_id) = &options.task_id {
            query.push(("task_id", task_id.clone()));
        }
        if let Some(tenant_id) = &options.tenant_id {
            query.push(("tenant_id", tenant_id.clone()));
        }
        if let Some(state) = &options.state {
            query.push(("state", state.to_string()));
        }
        if let Some(mode) = &options.mode {
            query.push(("mode", mode.to_string()));
        }
        if let Some(name) = &options.name {
            query.push(("name", name.clone()));
        }
        if let Some(layer_key) = &options.layer_key {
            query.push(("layer_key", layer_key.clone()));
        }
        if let Some(arch) = &options.arch {
            query.push(("arch", arch.to_string()));
        }
        pagination(&mut query, options.limit, &options.cursor);
        let request = self
            .t
            .request(Method::GET, &["v1", "snapshots"])
            .query(&query);
        self.t.send_json(request).await
    }

    /// Looks up the newest ready build artifact for one layer cache key and
    /// architecture, reporting `None` when there is none.
    ///
    /// A miss is not an error. A cold cache is the normal state of a first
    /// build, and modelling it as a failure would make every caller wrap
    /// this in error handling to discover nothing was wrong.
    ///
    /// `arch` is required rather than defaulted. An ext4 rootfs is not
    /// portable across architectures, so resolving without one could hand
    /// back an artifact the caller cannot boot, at the exact moment it
    /// believes it got a cache hit.
    ///
    /// The scope searched comes from how the client authenticated. There is
    /// deliberately no tenant parameter: a caller that could name its own
    /// scope could read another tenant's cache.
    pub async fn resolve(&self, layer_key: &str, arch: Arch) -> Result<Option<Snapshot>, Error> {
        require(layer_key, "layer key")?;

        // the found flag stays private so callers get Option instead, which
        // cannot be confused with a decode that happened to produce a
        // zero-value snapshot.
        #[derive(Deserialize)]
        struct Wire {
            #[serde(default)]
            found: bool,
            #[serde(default)]
            snapshot: Option<Snapshot>,
        }

        let request = self
            .t
            .request(Method::GET, &["v1", "snapshots", "resolve"])
            .query(&[("layer_key", layer_key), ("arch", arch.as_str())]);
        let wire: Wire = self.t.send_json(request).await?;
        Ok(if wire.found { wire.snapshot } else { None })
    }

    /// Returns one snapshot by id.
    pub async fn get(&self, snapshot_id: &str) -> Result<Snapshot, Error> {
        require(snapshot_id, "snapshot id")?;
        let request = self
            .t
            .request(Method::GET, &["v1", "snapshots", snapshot_id]);
        self.t.send_json(request).await
    }

    /// Deletes a snapshot.
    pub async fn delete(&self, snapshot_id: &str) -> Result<(), Error> {
        require(snapshot_id, "snapshot id")?;
        let request = self
            .t
            .request(Method::DELETE, &["v1", "snapshots", snapshot_id]);
        self.t.send_unit(request).await
    }

    /// Restores a snapshot into its environment.
    pub async fn restore(&self, snapshot_id: &str) -> Result<(), Error> {
        require(snapshot_id, "snapshot id")?;
        let request = self
            .t
            .request(Method::POST, &["v1", "snapshots", snapshot_id])
            .query(&[("action", "restore")]);
        self.t.send_unit(request).await
    }
}
